sdk/eventhubs/azure-messaging-eventhubs-checkpointstore-blob/src/main/java/com/azure/messaging/eventhubs/checkpointstore/blob/MetricsHelper.java

```diff
@@ -0,0 +1,187 @@
+import com.azure.core.util.metrics.MeterProvider;
+import com.azure.messaging.eventhubs.models.Checkpoint;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicReference;
+
+class MetricsHelper {
+    private static final ClientLogger LOGGER = new ClientLogger(MetricsHelper.class);
+
+    // Make sure attribute names are consistent across AMQP Core, EventHubs, ServiceBus when applicable
+    // and mapped correctly in OTel Metrics https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/core/azure-core-metrics-opentelemetry/src/main/java/com/azure/core/metrics/opentelemetry/OpenTelemetryAttributes.java
+    private static final String ENTITY_NAME_KEY = "entityName";
+    private static final String HOSTNAME_KEY = "hostName";
+    private static final String PARTITION_ID_KEY = "partitionId";
+    private static final String CONSUMER_GROUP_KEY = "consumerGroup";
+    private static final String STATUS_KEY = "status";
+
+    // since checkpoint store is stateless it might be used for endless number of eventhubs.
+    // we'll have as many subscriptions as there are combinations of fqdn, eventhub name, partitionId and consumer group.
+    // In the unlikely case it's shared across a lot of EH client instances, metrics would be too costly
+    // and unhelpful. So, let's just set a hard limit on number of subscriptions.
+    private static final int MAX_ATTRIBUTES_SETS = 100;
+
+    private static final String PROPERTIES_FILE = "azure-messaging-eventhubs-checkpointstore-blob.properties";
+    private static final String NAME_KEY = "name";
+    private static final String VERSION_KEY = "version";
+    private static final String LIBRARY_NAME;
+    private static final String LIBRARY_VERSION;
+    private static MeterProvider meterProvider = MeterProvider.getDefaultProvider();
+
+    private static final String UNKNOWN = "UNKNOWN";
+
+    static {
+        final Map<String, String> properties = CoreUtils.getProperties(PROPERTIES_FILE);
+        LIBRARY_NAME = properties.getOrDefault(NAME_KEY, UNKNOWN);
+        LIBRARY_VERSION = properties.getOrDefault(VERSION_KEY, UNKNOWN);
+    }
+
+    private final ConcurrentHashMap<String, TelemetryAttributes> common = new ConcurrentHashMap<>();
+    private final ConcurrentHashMap<String, TelemetryAttributes> checkpointFailure = new ConcurrentHashMap<>();
+    private final ConcurrentHashMap<String, TelemetryAttributes> checkpointSuccess = new ConcurrentHashMap<>();
+    private final ConcurrentHashMap<String, CurrentValue> seqNoSubscriptions = new ConcurrentHashMap<>();
+
+    private volatile boolean maxCapacityReached = false;
+
+    private final Meter meter;
+    private final LongGauge lastSequenceNumber;
+    private final LongCounter checkpointCounter;
+    private final boolean isEnabled;
+    MetricsHelper(MetricsOptions metricsOptions) {
+        if (metricsOptions != null && !metricsOptions.isEnabled()) {
```

Comment:
Is it possible for the options to change (i.e. reloaded configuration after creation?) If not, the `options.isEnabled()` and `null` checks are a lot to follow... I wondered if making just a `NoOpMetricsHelper` (they'd inherit from some interface) would make the logic easier to follow.

Reply:
options can't change, but I'd rather avoid adding noop class - it'd prevent JVM from inlining MeterHelper methods and would affect performance. I'll try to make it more readable though :)